packages/jni-swig-stub/src/main/jni/realm_api_helpers.cpp

```diff
@@ -0,0 +1,333 @@
+                                     on_change_method,
+                                     reinterpret_cast<jlong>(changes));
+            },
+            [](void *userdata,
+               const realm_async_error_t *async_error) {
+                // TODO Propagate errors to callback
+                //  https://github.com/realm/realm-kotlin/issues/303
+            },
+            // C-API currently uses the realm's default scheduler no matter what passed here
+            NULL
+    );
+}
+
+realm_notification_token_t *
+register_object_notification_cb(realm_object_t *object, jobject callback) {
+    auto jenv = get_env();
+    static jclass notification_class = jenv->FindClass("io/realm/interop/NotificationCallback");
+    static jmethodID on_change_method = jenv->GetMethodID(notification_class, "onChange", "(J)V");
+
+    return realm_object_add_notification_callback(
+            object,
+            // Use the callback as user data
+            static_cast<jobject>(get_env()->NewGlobalRef(callback)),
+            [](void *userdata) {
+                get_env(true)->DeleteGlobalRef(static_cast<jobject>(userdata));
+            },
+            // change callback
+            [](void *userdata, const realm_object_changes_t *changes) {
+                // TODO API-NOTIFICATION Consider catching errors and propagate to error callback
+                //  like the C-API error callback below
+                //  https://github.com/realm/realm-kotlin/issues/303
+                auto jenv = get_env(true);
+                if (jenv->ExceptionCheck()) {
+                    jenv->ExceptionDescribe();
+                    throw std::runtime_error("An unexpected Error was thrown from Java. See LogCat");
+                }
+                jenv->CallVoidMethod(static_cast<jobject>(userdata),
+                                     on_change_method,
+                                     reinterpret_cast<jlong>(changes));
+            },
+            [](void *userdata,
+               const realm_async_error_t *async_error) {
+                // TODO Propagate errors to callback
+                //  https://github.com/realm/realm-kotlin/issues/303
+            },
+            // C-API currently uses the realm's default scheduler no matter what passed here
+            NULL
+    );
+}
+
+void register_login_cb(realm_app_t* app, realm_app_credentials_t* credentials, jobject callback) {
+    auto jenv = get_env();
+    static jclass notification_class = jenv->FindClass("io/realm/interop/OperationCallback");
+    static jmethodID on_success_method = jenv->GetMethodID(notification_class, "onSuccess", "(Lio/realm/interop/NativePointer;)V");
+    static jmethodID on_error_method = jenv->GetMethodID(notification_class, "onError", "(Ljava/lang/Throwable;)V");
+
+    realm_app_log_in_with_credentials(
+            app,
+            credentials,
+            [](void* userdata, realm_user_t* user, realm_error_t* error) {
```

Comment:
These callbacks are shared between a lot of API calls. In java we have helper methods for creating these:

https://github.com/realm/realm-java/blob/master/realm/realm-library/src/main/cpp/java_network_transport.hpp#L95

https://github.com/realm/realm-java/blob/master/realm/realm-library/src/main/cpp/java_network_transport.hpp#L119

Callsite, then becomes much less code:

https://github.com/realm/realm-java/blob/master/realm/realm-library/src/main/cpp/io_realm_internal_objectstore_OsApp.cpp#L192

We should do something similar, but just adding FIXME for now is fine.

Reply:
Adding a FIXME referencing the Java implementation. Makes most sense to do when introducing the next call that follows this pattern.